components/brave_wallet/browser/json_rpc_service.cc

```diff
@@ -1208,66 +1211,73 @@ void JsonRpcService::OnEnsGetEthAddr(
-    const std::string& chain_id,
+void JsonRpcService::UnstoppableDomainsResolveDns(
     const std::string& domain,
-    const std::vector<std::string>& keys,
-    UnstoppableDomainsProxyReaderGetManyCallback callback) {
-  const std::string contract_address =
-      GetUnstoppableDomainsProxyReaderContractAddress(chain_id);
-  if (contract_address.empty()) {
-    std::move(callback).Run(
-        std::vector<std::string>(), mojom::ProviderError::kInvalidParams,
-        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    UnstoppableDomainsResolveDnsCallback callback) {
+  if (ud_resolve_dns_calls_->HasCall(domain)) {
+    ud_resolve_dns_calls_->AddCallback(domain, std::move(callback));
     return;
   }
 
-  auto data = unstoppable_domains::GetMany(keys, domain);
-  if (!data) {
+  if (!IsValidUnstoppableDomain(domain)) {
     std::move(callback).Run(
-        std::vector<std::string>(), mojom::ProviderError::kInvalidParams,
+        GURL(), mojom::ProviderError::kInvalidParams,
         l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
     return;
   }
 
-  GURL network_url = GetNetworkURL(prefs_, chain_id, mojom::CoinType::ETH);
-  if (!network_url.is_valid()) {
+  auto data = unstoppable_domains::GetMany(unstoppable_domains::GetRecordKeys(),
+                                           domain);
+  if (!data) {
     std::move(callback).Run(
-        std::vector<std::string>(), mojom::ProviderError::kInvalidParams,
+        GURL(), mojom::ProviderError::kInvalidParams,
         l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
     return;
   }
 
-  auto internal_callback =
-      base::BindOnce(&JsonRpcService::OnUnstoppableDomainsProxyReaderGetMany,
-                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
-  RequestInternal(
-      eth::eth_call("", contract_address, "", "", "", *data, "latest"), true,
-      network_url, std::move(internal_callback));
+  ud_resolve_dns_calls_->AddCallback(domain, std::move(callback));
+  for (const auto& chain_id : ud_resolve_dns_calls_->GetChains()) {
+    auto internal_callback =
+        base::BindOnce(&JsonRpcService::OnUnstoppableDomainsResolveDns,
+                       weak_ptr_factory_.GetWeakPtr(), domain, chain_id);
+    auto eth_call = eth::eth_call(
+        "", GetUnstoppableDomainsProxyReaderContractAddress(chain_id), "", "",
+        "", *data, "latest");
+    RequestInternal(std::move(eth_call), true,
+                    GetUnstoppableDomainsRpcUrl(chain_id),
+                    std::move(internal_callback));
+  }
 }
 
-void JsonRpcService::OnUnstoppableDomainsProxyReaderGetMany(
-    UnstoppableDomainsProxyReaderGetManyCallback callback,
+void JsonRpcService::OnUnstoppableDomainsResolveDns(
+    const std::string& domain,
+    const std::string& chain_id,
     const int status,
     const std::string& body,
     const base::flat_map<std::string, std::string>& headers) {
   if (status < 200 || status > 299) {
```

Comment:
Ok ideally we'd want to use a 200 here unless there was an expected reason to handle another 200 class HTTP response. This is to have more defensive programming logic in case the Polygon gateway changes underneath us. This isn't a blocking change though and I'll leave it up to you whether or not you think it's necessary to change.